sharding/sharding.go

```diff
@@ -0,0 +1,311 @@
+		}
+		shardsNum := float64(clientCfg.ShardsCount) * float64(clientCluster.weight) / float64(weightSum)
+		for i := offset; i < offset+int(shardsNum); i++ {
+			shardName := fmt.Sprintf("%s-%d", clientCfg.Name, i)
+			shardClusterMap[shardName] = clientCluster
+		}
+		offset += int(shardsNum)
+	}
+	return shardClusterMap, nil
+}
+
+func (rf ringFactory) uniqBackends(clientCfg config.ClientConfig) ([]*url.URL, error) {
+	allBackendsSet := make(map[config.YAMLURL]bool)
+	for _, name := range clientCfg.Clusters {
+		clientCluster, err := rf.getCluster(name)
+		if err != nil {
+			return nil, err
+		}
+		for _, backendURL := range clientCluster.backends {
+			allBackendsSet[backendURL] = true
+		}
+	}
+	var uniqBackendsSlice []*url.URL
+	for url := range allBackendsSet {
+		uniqBackendsSlice = append(uniqBackendsSlice, url.URL)
+	}
+	return uniqBackendsSlice, nil
+}
+
+func (rf ringFactory) clientRing(clientCfg config.ClientConfig) (shardsRing, error) {
+	weightSum := uint(0)
+	clientClusters := make([]cluster, 0, len(clientCfg.Clusters))
+	regressionMap := make(map[string]cluster)
+	var previousCluster cluster
+	for i, name := range clientCfg.Clusters {
+		clientCluster, err := rf.getCluster(name)
+		if err != nil {
+			return shardsRing{}, err
+		}
+
+		weightSum += clientCluster.weight
+		clientClusters = append(clientClusters, clientCluster)
+
+		if i > 0 {
+			regressionMap[name] = previousCluster
+		}
+		previousCluster = clientCluster
+	}
+	shardMap, err := rf.mapShards(weightSum, clientCfg)
+	if err != nil {
+		return shardsRing{}, err
+	}
+	cHashMap := consistenthash.New(1, nil)
+	for shardID := range shardMap {
+		cHashMap.Add(shardID)
+	}
+
+	allBackendsSlice, err := rf.uniqBackends(clientCfg)
+	if err != nil {
+		return shardsRing{}, err
+	}
+	allBackendsRoundTripper := transport.NewMultiTransport(
+		rf.transport,
+		allBackendsSlice,
+		rf.multipleResponseHandler)
+
+	return shardsRing{cHashMap, shardMap, allBackendsRoundTripper, regressionMap, rf.conf.ClusterSyncLog}, nil
+}
+
+func newRingFactory(conf config.Config, transport http.RoundTripper, respHandler transport.MultipleResponsesHandler) ringFactory {
+	return ringFactory{
+		conf:                    conf,
+		transport:               transport,
+		multipleResponseHandler: respHandler,
+		clusters:                make(map[string]cluster),
+	}
+}
+
+// NewHandler constructs http.Handler
+func NewHandler(conf config.Config) (http.Handler, error) {
+	// clustersMap, _ := mapClusterTypes(conf)
+	clustersNames := make([]string, 0, len(conf.Clusters))
+	for name := range conf.Clusters {
+		clustersNames = append(clustersNames, name)
+	}
+
+	conf.Mainlog.Printf("Configured clusters: %s", strings.Join(clustersNames, ", "))
+
+	httptransp, err := httphandler.ConfigureHTTPTransport(conf)
+	if err != nil {
+		return nil, err
+	}
+	respHandler := httphandler.NewMultipleResponseHandler(conf)
+	rings := newRingFactory(conf, httptransp, respHandler)
+	// TODO: Multiple clients
```

Comment:
Add more contenxt to this todo. Issue number or detailed comment what should be done.